From de2e5171433126d340573cb7d0d4fcac084ab2a0 Mon Sep 17 00:00:00 2001
From: Kefeng Wang <wangkefeng.wang@huawei.com>
Date: Mon, 5 Dec 2022 22:03:27 +0800
Subject: [PATCH] mm: add cond_resched() in swapin_walk_pmd_entry()
Git-commit: de2e5171433126d340573cb7d0d4fcac084ab2a0
Patch-mainline: v6.2-rc1
References: bsn#12
Modified-by-SEL: Yes, modified due to different contexy


When handling MADV_WILLNEED in madvise(), a soflockup may occurr in
swapin_walk_pmd_entry() if swapping in lots of memory on a slow device.
Add a cond_resched() to avoid the possible softlockup.

Link: https://lkml.kernel.org/r/20221205140327.72304-1-wangkefeng.wang@huawei.com
Fixes: 1998cc048901 ("mm: make madvise(MADV_WILLNEED) support swap file prefetch")
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Shaohua Li <shli@fusionio.com>
Cc: Hugh Dickins <hughd@google.com>
Cc: Rik van Riel <riel@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 mm/madvise.c |    1 +
 1 file changed, 1 insertion(+)

--- a/mm/madvise.c
+++ b/mm/madvise.c
@@ -213,6 +213,7 @@ static int swapin_walk_pmd_entry(pmd_t *
 		if (page)
 			put_page(page);
 	}
+	cond_resched();
 
 	return 0;
 }
